apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Utilities/EqualityHelpers.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+
+namespace Google.Cloud.Firestore
+{
+    /// <summary>
+    /// Methods to help implement IEquatable{T}.
+    /// </summary>
+    internal static class EqualityHelpers
+    {
+        private const int HashInitialValue = 3581;
+
+        /// <summary>
+        /// Checks if two lists are equal.
+        /// </summary>
+        internal static bool ListsEqual<T>(IReadOnlyList<T> left, IReadOnlyList<T> right)
+            where T : IEquatable<T>
+        {
+            if (ReferenceEquals(left, right))
+            {
+                return true;
+            }
+            if (ReferenceEquals(left, null) || ReferenceEquals(right, null))
+            {
+                return false;
+            }
+            if (left.Count != right.Count)
+            {
+                return false;
+            }
+            for (int i = 0; i < left.Count; i++)
+            {
+                if (!left[i].Equals(right[i]))
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Computes a hash code for a list
+        /// </summary>
+        internal static int GetListHashCode<T>(IReadOnlyList<T> list) where T : IEquatable<T>
+        {            
+            if (list == null)
+            {
+                return 0;
+            }
+            unchecked
```

Comment:
nit: unchecked is the default, so is there a need to explicitly use it?

Reply:
While it's the default, I suspect we may well be tempted to make checked the default at some point... but hash combinations should *never* be done in a checked context.